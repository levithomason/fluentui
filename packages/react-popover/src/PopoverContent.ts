export * from './components/PopoverContent/index';
