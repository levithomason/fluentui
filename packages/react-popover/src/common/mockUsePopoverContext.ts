import { PopoverContextValue, usePopoverContext } from '../popoverContext';

/**
 * A test utility to mock the usePopoverContext hook that relies on context selector
 * Don't forget to call `jest.mock(**\/popupContext.ts)` while using this
 * @param options Popover context values to set for testing
 */
export const mockPopoverContext = (options: Partial<PopoverContextValue> = {}) => {
  const mockContext: PopoverContextValue = {
    target: null,
    triggerRef: { current: null },
    contentRef: { current: null },
    open: false,
    setOpen: jest.fn(),
    mountNode: null,
    openOnContext: false,
    openOnHover: false,
    ...options,
  };
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (usePopoverContext as jest.Mock).mockImplementation((selector: (context: PopoverContextValue) => any) => {
    return selector(mockContext);
  });
};
