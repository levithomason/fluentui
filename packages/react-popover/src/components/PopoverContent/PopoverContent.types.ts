import * as React from 'react';
import { ComponentProps, ComponentState } from '@fluentui/react-utilities';
import { PopoverContextValue } from '../../popoverContext';

/**
 * PopoverContent Props
 */
export interface PopoverContentProps extends ComponentProps, React.HTMLAttributes<HTMLElement> {}

/**
 * Names of the shorthand properties in PopoverContentProps
 */
export type PopoverContentShorthandProps = never;

/**
 * Names of PopoverContentProps that have a default value in usePopoverContent
 */
export type PopoverContentDefaultedProps = never;

/**
 * PopoverContent State
 */
export interface PopoverContentState
  extends ComponentState<PopoverContentProps, PopoverContentShorthandProps, PopoverContentDefaultedProps>,
    Pick<PopoverContextValue, 'open' | 'mountNode' | 'noArrow' | 'size' | 'brand' | 'inverted'> {
  ref: React.Ref<HTMLElement>;
  /**
   * Ref to the arrow element
   */
  arrowRef?: React.Ref<HTMLDivElement>;

  /**
   * CSS class for the arrow element
   */
  arrowClassName?: string;
}
