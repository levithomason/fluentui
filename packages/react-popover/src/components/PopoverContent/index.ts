export * from './PopoverContent';
export * from './PopoverContent.types';
export * from './renderPopoverContent';
export * from './usePopoverContent';
export * from './usePopoverContentStyles';
