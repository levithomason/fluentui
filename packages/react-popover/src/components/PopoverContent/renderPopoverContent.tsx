import * as React from 'react';
import { getSlots } from '@fluentui/react-utilities';
import { Portal } from '@fluentui/react-portal';
import { PopoverContentState } from './PopoverContent.types';

/**
 * Render the final JSX of PopoverContent
 */
export const renderPopoverContent = (state: PopoverContentState) => {
  const { slots, slotProps } = getSlots(state);

  // TODO should hidden Popovers be supported ?
  if (!state.open) {
    return null;
  }

  return (
    <Portal mountNode={state.mountNode}>
      <slots.root {...slotProps.root}>
        {!state.noArrow && <div ref={state.arrowRef} className={state.arrowClassName} />}
        {state.children}
      </slots.root>
    </Portal>
  );
};
