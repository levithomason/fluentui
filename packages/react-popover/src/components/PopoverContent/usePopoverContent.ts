import * as React from 'react';
import { makeMergeProps, useMergedRefs } from '@fluentui/react-utilities';
import { useFocusFinders } from '@fluentui/react-tabster';
import { PopoverContentProps, PopoverContentState } from './PopoverContent.types';
import { usePopoverContext } from '../../popoverContext';

const mergeProps = makeMergeProps<PopoverContentState>({});

/**
 * Create the state required to render PopoverContent.
 *
 * The returned state can be modified with hooks such as usePopoverContentStyles,
 * before being passed to renderPopoverContent.
 *
 * @param props - props from this instance of PopoverContent
 * @param ref - reference to root HTMLElement of PopoverContent
 * @param defaultProps - (optional) default prop values provided by the implementing type
 */
export const usePopoverContent = (
  props: PopoverContentProps,
  ref: React.Ref<HTMLElement>,
  defaultProps?: PopoverContentProps,
): PopoverContentState => {
  const contentRef = usePopoverContext(context => context.contentRef);
  const open = usePopoverContext(context => context.open);
  const openOnHover = usePopoverContext(context => context.openOnHover);
  const setOpen = usePopoverContext(context => context.setOpen);
  const mountNode = usePopoverContext(context => context.mountNode);
  const arrowRef = usePopoverContext(context => context.arrowRef);
  const size = usePopoverContext(context => context.size);
  const noArrow = usePopoverContext(context => context.noArrow);
  const brand = usePopoverContext(context => context.brand);
  const inverted = usePopoverContext(context => context.inverted);

  const state = mergeProps(
    {
      brand,
      inverted,
      noArrow,
      size,
      arrowRef,
      open,
      mountNode,
      role: 'dialog',
      ref: useMergedRefs(ref, contentRef),
    },
    defaultProps,
    props,
  );

  const {
    onMouseEnter: onMouseEnterOriginal,
    onMouseLeave: onMouseLeaveOriginal,
    onKeyDown: onKeyDownOriginal,
  } = state;
  state.onMouseEnter = (e: React.MouseEvent<HTMLElement>) => {
    if (openOnHover) {
      setOpen(e, true);
    }

    onMouseEnterOriginal?.(e);
  };

  state.onMouseLeave = (e: React.MouseEvent<HTMLElement>) => {
    if (openOnHover) {
      setOpen(e, false);
    }

    onMouseLeaveOriginal?.(e);
  };

  state.onKeyDown = (e: React.KeyboardEvent<HTMLElement>) => {
    // only close if the event happened inside the current popover
    // If using a stack of inline popovers, the user should call `stopPropagation` to avoid dismissing the entire stack
    if (e.key === 'Escape' && contentRef.current?.contains(e.target as HTMLElement)) {
      setOpen(e, false);
    }

    onKeyDownOriginal?.(e);
  };

  const { findFirstFocusable } = useFocusFinders();

  // TODO Temporary, use tabster modalizer for a real focus trap
  React.useEffect(() => {
    if (state.open && contentRef.current) {
      const firstFocusable = findFirstFocusable(contentRef.current);
      firstFocusable?.focus();
    }
  }, [contentRef, findFirstFocusable, state.open]);
  return state;
};
