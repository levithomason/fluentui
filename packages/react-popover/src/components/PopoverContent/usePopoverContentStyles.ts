import { makeStyles, mergeClasses } from '@fluentui/react-make-styles';
import { PopoverContentState } from './PopoverContent.types';

/**
 * Styles for the root slot
 */
const useStyles = makeStyles({
  root: theme => ({
    backgroundColor: theme.alias.color.neutral.neutralBackground1,
    boxShadow: theme.alias.shadow.shadow16,
  }),
});

/**
 * Apply styling to the PopoverContent slots based on the state
 */
export const usePopoverContentStyles = (state: PopoverContentState): PopoverContentState => {
  const styles = useStyles();
  state.className = mergeClasses(styles.root, state.className);

  return state;
};
