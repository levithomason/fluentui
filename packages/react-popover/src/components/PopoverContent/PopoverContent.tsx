import * as React from 'react';
import { usePopoverContent } from './usePopoverContent';
import { PopoverContentProps } from './PopoverContent.types';
import { renderPopoverContent } from './renderPopoverContent';
import { usePopoverContentStyles } from './usePopoverContentStyles';

/**
 * PopoverContent component renders react children in a positioned box
 */
export const PopoverContent = React.forwardRef<HTMLElement, PopoverContentProps>((props, ref) => {
  const state = usePopoverContent(props, ref);

  usePopoverContentStyles(state);
  return renderPopoverContent(state);
});

PopoverContent.displayName = 'PopoverContent';
