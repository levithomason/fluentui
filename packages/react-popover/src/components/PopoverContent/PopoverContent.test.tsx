import * as React from 'react';
import { PopoverContent } from './PopoverContent';
import { render, fireEvent } from '@testing-library/react';
import { ReactWrapper } from 'enzyme';
import { isConformant } from '../../common/isConformant';
import { Portal } from '@fluentui/react-portal';
import { mockPopoverContext } from '../../common/mockUsePopoverContext';
import { PopoverContentProps } from './PopoverContent.types';

jest.mock('../../popoverContext');

describe('PopoverContent', () => {
  isConformant({
    // as render test will pass a span tag which is also considered one of the skipped helperComponents
    disabledTests: ['as-renders-html'],
    Component: PopoverContent,
    displayName: 'PopoverContent',
    requiredProps: { open: true } as PopoverContentProps,
    helperComponents: [Portal, 'span'],
  });

  let wrapper: ReactWrapper | undefined;

  afterEach(() => {
    if (wrapper) {
      wrapper.unmount();
      wrapper = undefined;
    }
  });

  beforeEach(() => {
    mockPopoverContext({ open: true });
  });

  // PopoverContent is rendered by a Portal so won't be available in the rendered container
  const testid = 'component';

  /**
   * Note: see more visual regression tests for PopoverContent in /apps/vr-tests.
   */
  it('renders a default state', () => {
    const { getByTestId } = render(<PopoverContent data-testid={testid}>Default PopoverContent</PopoverContent>);
    expect(getByTestId(testid)).toMatchSnapshot();
  });

  it.each([
    ['onMouseEnter', fireEvent.mouseEnter],
    ['onMouseLeave', fireEvent.mouseLeave],
    ['onKeyDown', fireEvent.keyDown],
  ])('should keep the original %s handler', (handler, triggerEvent) => {
    // Arrange
    const spy = jest.fn();
    const { getByTestId } = render(
      <PopoverContent data-testid={testid} {...{ [handler]: spy }}>
        Content
      </PopoverContent>,
    );

    // Act
    triggerEvent(getByTestId(testid));

    // Assert
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('should not render when open is false', () => {
    // Arrange
    mockPopoverContext({ open: false });
    const { queryByTestId } = render(<PopoverContent data-testid={testid}>Content</PopoverContent>);

    // Assert
    expect(queryByTestId(testid)).toBeNull();
  });
});
