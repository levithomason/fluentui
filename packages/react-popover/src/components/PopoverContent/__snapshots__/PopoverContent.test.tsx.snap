// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopoverContent renders a default state 1`] = `
<div
  class=""
  data-testid="component"
  role="dialog"
>
  <div
    class=""
  />
  Default PopoverContent
</div>
`;
